### PR TITLE
fix(metrics): use current actor id for backpressure

### PR DESCRIPTION
**Pull request by skyzh**
Signed-off-by: Alex Chi <iskyzh@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

As title. The back pressure rate should be normal now.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/stream/src/executor/dispatch.rs [hunk 1]
```diff
@@ -228,14 +180,23 @@ impl DispatchExecutorInner {
                         dispatcher.dispatch_data(chunk.clone()).await?;
                     }
                 }
+                self.metrics
+                    .actor_output_buffer_blocking_duration_ns
+                    .with_label_values(&[&self.actor_id_str])
+                    .inc_by(start_time.elapsed().as_nanos() as u64);
```
Comment (Sunt-ing) [inline src/stream/src/executor/dispatch.rs:186]: So it will not only count the blocking time but also the normal dispatch time?

Comment (skyzh) [reply src/stream/src/executor/dispatch.rs:186]: If there's no back pressure, the time should be close to 0.

Comment (Sunt-ing) [reply src/stream/src/executor/dispatch.rs:186]: Have you benched this?

Comment (Sunt-ing) [reply src/stream/src/executor/dispatch.rs:186]: I think you also refactored the Grafana dashboard in this PR, so maybe this PR can be merged after my PR is merged?

**Comment by skyzh** [comment]
image

File: src/stream/src/executor/dispatch.rs (hunk 1 above)
Comment (skyzh) [reply src/stream/src/executor/dispatch.rs:186]: > If there is actually no backpressure and resource is over-provision, but the metrics show that there is, the metrics can be pretty misleading.

Is there any workload without back-pressure?

**Comment by Sunt-ing** [comment]
The metrics in this PR maybe never be 0 even when the system is actually pretty over-provision. This will make our scaling-in behave rather weird.

**Review by BugenZhao — APPROVED** [review]
LGTM

**Comment by skyzh** [comment]
> The metrics in this PR maybe never be 0 even when the system is actually pretty over-provision. This will make our scaling-in behave rather weird.

You can first come up with a good scaling policy, then see whether this will affect your policy.

**Comment by Sunt-ing** [comment]
@BugenZhao May you explain your behavior?

**Comment by BugenZhao** [comment]
> @BugenZhao May you explain your behavior?

> The metrics in this PR maybe never be 0 even when the system is actually pretty over-provision.

Under original implementation, the `capacity` will be stale immediately after the `if` statement, so it's still possible that the queue is not full, and we count the "normal dispatch time". Therefore, we should never rely on the fact that "the back pressure rate is exactly zero" for scaling policy, unless we measure the time of `next` future polling. So this does not make much sense.

For PR merging, I was asked to review this PR and had some offline discussions with the author, while not realizing #3723. Based on the above statements, I don't think there's any big problem with this fix, so I approved this. Anything objection to this?

**Comment by Sunt-ing** [comment]
> the capacity will be stale immediately after the if statement
This possibility is extremely low, so definitely we can accept.

**Comment by BugenZhao** [comment]
> 

Cool. So I believe the approach in this PR is also acceptable.

**Comment by Sunt-ing** [comment]
In my implementation https://github.com/singularity-data/risingwave/pull/3723, only when this "extremely low possibility" is hit, the overhead would equal the overhead in this PR.

**Comment by skyzh** [comment]
> The metrics in this PR maybe never be 0 even when the system is actually pretty over-provision. This will make our scaling-in behave rather weird.

If your scaling policy is so sensitive to such small error values, then it's not robust enough, and should never be used by our users. You should redesign your scaling policy, instead of being peculiar about this fix. You should feel fortunate that "an actor is backpressured 4 seconds every 1 second" bug is found sooner than months later you've already based your policy on such wrong metrics.

> In my implementation https://github.com/singularity-data/risingwave/pull/3723, only when this "extremely low possibility" is hit, the overhead would equal the overhead in this PR.

Being peculiar about such edge cases doesn't make any sense. Shall we consider the case that the CPU clock speed is slightly different on each core? Shall we consider the time for holding the mutex inside channel.send? If you're so addicted to such accuracy, I can raise millions of things that you'll need to consider. Thus, the principle is to make every thing simple and easy to understand.

> @BugenZhao May you explain your behavior?

I believe our discussions have been settled until this point. It's you that are making our normal development process uncomfortable. I have no idea why you've got such big reaction to this PR. I believe I've fully taken your review comments into consideration before I think it's ready to merge this PR.

You should think about yourself, Ting, instead of blaming us that our code doesn't fit your need. Doing an engineering work is harder than you think. It's impossible to make everyone happy. You can just come up with some metrics that "scaling service" will use but nobody understands what it means. You're happy enough to make your own things work, but other developers will feel confused that why we present the "de/ser" duration in such a way on the Grafana dashboard. RisingWave is not a project owned by you, or just for the serverless scaling research, also not by me. We all should make it go in the direction that most people will feel comfortable with and fit our products' need.

Regarding your questions in the Slack channel,

> how to prevent this evil approve+merge action? shall we set some regulations?

For (1), You cannot judge whether a commit is evil or not based on your own preference. As I've said, it's impossible to make everyone happy in a real project. If you disagrees on a PR, you should propose a better solution instead of blaming that we're doing evil things. You should think of why your proposed changes are necessary and whether it's good to have it.

For (2), the standard is always one approval and passing CI tests to get some patches merged. If the criteria is met, then we can get PRs merged. Of course, someone might be opposite to some changes, so do I. For me, I'll propose better solutions or fix it by myself. No developers will ask for "explanations" on the merge/reject action itself -- it's like asking why I'm getting up late at morning, and the answer is simple -- I'm okay with it.

Scaling our system is important of course, but it's not the only important thing. I'm not less exciting than you to have real-time scale in our system, but we still have a lot of things to do before we can tell whether a scale policy is good or bad. From my perspective, you're hurrying in some changes without enough thinking. There're so many on-going prerequisite works about scaling, like actor migration, vnode partition, performance investigation, etc. At the same time, we have the same code quality requirement for the scaling project as other ongoing projects. You should learn about them, instead of just telling whether some 0.0001% flatuation will affect your scaling policy or not. Your research depends on many works on the kernel side, and you should contribute code with effort instead of just making your research works.

Please keep that in mind, and hope that we can both help make the project better.

**Pull request merged**